File: Leetcode/TypeScript/108. Convert Sorted Array to Binary Search Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function sortedArrayToBST(nums: number[]): TreeNode | null {
    if(!nums.length) return null;
    
    let mid: number = Math.floor(nums.length/2);
    let left: number[] = nums.slice(0,mid);
    let right: number[] = nums.slice(mid+1, nums.length);
    
    return new TreeNode(nums[mid], sortedArrayToBST(left), sortedArrayToBST(right))
};

File: Leetcode/C++/108. Convert Sorted Array to Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return arrayToBST(nums, 0, nums.size()-1);
    }
    
    TreeNode* arrayToBST(vector<int>& nums, int s, int e) {
        if(e < s) return nullptr;
        int m = (s+e)/2;
        
        auto r = new TreeNode(nums[m]);
        r->left = arrayToBST(nums, s, m-1);
        r->right = arrayToBST(nums, m+1, e);
        
        return r;
    }
};
